go/apps/mixnet/proxy.go

```diff
@@ -205,3 +228,145 @@ func (p *ProxyContext) nextID() (id uint64) {
 	p.id++
 	return id
 }
+
+// Accept partially implements the server role in version 5 of the SOCKS
+// protocol specified in RFC 1928. In particular, it only supports TCP clients
+// with no authentication who request CONNECT to IPv4 addresses; neither BIND
+// nor UDP ASSOCIATE are supported. Return the connection to the client and the
+// requested destination address.
+func (p *ProxyContext) Accept() (net.Conn, string, error) {
+	c, err := p.listener.Accept()
+	if err != nil {
+		return nil, "", err
+	}
+
+	// First, wait for greeting from client containing the SOCKS version and
+	// requested methods.
+	msg := make([]byte, MaxMsgBytes)
+	reply := make([]byte, MaxMsgBytes)
+	if _, err = c.Read(msg); err != nil {
+		c.Close()
+		return nil, "", err
+	}
+
+	// Only the NO AUTHENTICATION REQUIRED method is allowed. Note that this
+	// makes the server non-compliant since GSSAPI is not allowed.
+	ver := int(msg[0])
+	nmethods := int(msg[1])
+	ok := false
+	for _, method := range msg[2 : 2+nmethods] {
+		if method == 0x00 {
+			ok = true
+			break
+		}
+	}
+
+	// Second, reply with selected method.
+	reply[0] = SocksVersion
+	if ver == SocksVersion && ok {
+		reply[1] = 0x00 // NO AUTHENTICATION REQUIRED
+	} else {
+		reply[1] = 0xff // NO ACCEPTABLE METHODS
+	}
+
+	if _, err = c.Write(reply[:2]); err != nil {
+		c.Close()
+		return nil, "", err
+	}
+
+	// If NO ACCEPTABLE METHOD, the client closes the connection.
+	if reply[1] == 0xff {
+		c.Close()
+		return nil, "", errors.New("socks: client did not provide acceptable method")
+	}
+
+	// Third, wait for command from client.
+	bytes, err := c.Read(msg)
+	if err != nil {
+		c.Close()
+		return nil, "", err
+	}
+	ver = int(msg[0])
+	cmd := msg[1]
+	// msg[2] is a reserved byte in the protocol.
+	atyp := msg[3]
+
+	// Only CONNECT to IPv4 addresses is allowed. Since traffic will be proxied
+	// over the mixnet, don't connect to the intended host just yet. Reply to
+	// the client.
+	copy(reply, msg[:bytes])
+	if ver == SocksVersion && cmd == 0x01 /* CONNECT */ && atyp == 0x01 /* IPv4 */ {
+		reply[1] = 0x00 // SUCCEEDED.
+	} else {
+		reply[1] = 0x07 // COMMAND NOT SUPPORTED
+	}
+	if _, err = c.Write(reply[:bytes]); err != nil {
```

Comment:
Would it be possible to pull this Accept out into a socks_server.go file? It could implement net.Listener, and it would be of independent interest.